packages/iframe-stamper/src/index.ts

```diff
@@ -145,36 +155,58 @@ export class IframeStamper {
 
     // This is populated once the iframe is ready. Call `.init()` to kick off DOM insertion!
     this.iframePublicKey = null;
+
+    /**
+     * The MessageChannel API is used to establish secure communication between the parent page and the iframe.
+     * See https://developer.mozilla.org/en-US/docs/Web/API/MessageChannel
+     */
+    this.messageChannel = new MessageChannel();
   }
 
   /**
    * Inserts the iframe on the page and returns a promise resolving to the iframe's public key
    */
   async init(): Promise<string> {
     this.container.appendChild(this.iframe);
-    return new Promise((resolve, _reject) => {
-      window.addEventListener(
-        "message",
-        (event) => {
-          if (event.origin !== this.iframeOrigin) {
-            // There might be other things going on in the window, for example: react dev tools, other extensions, etc.
-            // Instead of erroring out
-            return;
-          }
-          if (event.data?.type === IframeEventType.PublicKeyReady) {
-            this.iframePublicKey = event.data["value"];
-            resolve(event.data["value"]);
-          }
-        },
-        false
+    /**
+     * Once the iframe is loaded, we send a message to the iframe to hand over the
+     * MessageChannel's second port, port2, and establish the secure communication channel.
+     * The iframe will use this port to send messages back to the parent page.
+     * See https://developer.mozilla.org/en-US/docs/Web/API/MessagePort/postMessage#transfer
+     */
+    this.iframe.addEventListener("load", () => {
+      if (
+        !this.iframe.contentWindow ||
+        !this.iframe.contentWindow.postMessage
+      ) {
+        throw new Error(
+          "contentWindow or contentWindow.postMessage does not exist"
```

Comment:
is there a scenario where we'd want granularity over this? in other words, is it particularly weird if `.postMessage` specifically doesn't exist?

Reply:
So TypeScript complains if you do `this.iframe.contentWindow.postMessage(...)` because the contentWindow is possibly null. I'm not really aware of a scenario where this would actually be the case in the wild unless there was an underlying bug in the browser itself. `contentWindow` is read-only so you wouldn't be able to null it out if you tried. 

This check is more of a defensive coding pattern in case something is really broken.

Reply:
100% agree with being defensive here. was just curious. i'm guessing having this in the wild is a pretty rare occurrence

Reply:
Yeah I think the internet would be broken